src/server/routes/search.rs

```diff
@@ -237,23 +226,21 @@ fn is_match_from_filter_list(
     Ok(false)
 }
 
-/// A helper function which returns the safe search level based on the url params
-/// and cookie value.
+/// A helper function to modify the safe search level based on the url params.
+/// The `safe_search` is the one in the user's cookie or
+/// the default set by the server config if the cookie was missing.
 ///
 /// # Argurments
 ///
-/// * `safe_search` - Safe search level from the url.
-/// * `cookie` - User's cookie
-/// * `default` - Safe search level to fall back to
-fn get_safesearch_level(safe_search: &Option<u8>, cookie: &Option<u8>, default: u8) -> u8 {
-    match safe_search {
-        Some(ss) => {
-            if *ss >= 3 {
-                default
-            } else {
-                *ss
-            }
+/// * `url_level` - Safe search level from the url.
+/// * `safe_search` - User's cookie, or the safe search level set by the server
+/// * `config_level` - Safe search level to fall back to
+fn get_safesearch_level(safe_search: &mut u8, url_level: &Option<u8>, config_level: u8) {
+    if let Some(search_level) = url_level {
+        if *search_level >= 3 {
+            *safe_search = config_level
+        } else {
+            *safe_search = *search_level;
```

Comment:
I would also suggest adding a fall back to cookie safe search level as well because you see there is a scenario where the engine will be configured by the user (server admin) using the config. Where the user might configure it with the safe search level of something other than to 3 or 4. Maybe because the user might not want to have a strictly SFW only server. In that case, if a user (user of the server) were to use such a server. And if let's say they don't like the default safe search level provided by the server admin of 2 or 0 and let's say they prefer 1. Then the user will have to each time set the safe search level from the search URL parameter, which you know is not great. Now if the user were to set it using the settings page which saves the safe search level as a cookie, this allows the user to set the safe search level only once. Hence, they will not need to change it again and again via the search URL parameter like they can still change it but, they might do it less often and prefer the more convenient settings page option. 

So, my suggestion would be to have a fallback in the else section. Where you can have a check for the none case for `search_level` variable, so when it is none it will fall back to using the cookie value if the user has set it. Thus overriding the value set by the server admin using the config. Otherwise, if it has not been set by the user then you know obviously you will fall back to the config value because the value set by the server admin was not overridden. So we will use it then. Also, if the `search_level` was provided using the search parameter, then we pass it as it is because it will override both the cookie and the value set by the server admin. I hope you got my idea :slightly_smiling_face: .